File: app/src/main/java/com/scruge/scruge/dependencies/navigation/NavigationController.java
package com.scruge.scruge.dependencies.navigation;

import android.transition.Slide;
import android.view.Gravity;
import android.view.View;

import com.scruge.scruge.view.views.NavigationBar;

import org.jetbrains.annotations.NotNull;

import java.util.Stack;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

public class NavigationController implements NavigationBar.Delegate {

    private enum Update { will, did }

    /// return true if handled, false to pass further
    public interface OnBackPressedListener {
        boolean onBackPressedHandled();
    }

    // FRAGMENT

    private boolean shouldHideNavigationBar = true;
    private NavigationBar navigationBar;
    private FragmentManager manager;
    private Stack<Fragment> fragmentStack = new Stack<>();
    private Fragment currentFragment;
    private int containerId;
    public boolean isVisible = true;

    public NavigationController(FragmentManager fragmentManager, int containerId) {
        this.containerId = containerId;
        this.manager = fragmentManager;
    }

    public int getContainerId() {
        return containerId;
    }

    public void replaceWith(final Fragment fragment) {
        final Fragment oldFragment = currentFragment;
        currentFragment = fragment;

        manager.executePendingTransactions();
        fragmentStack = new Stack<>();
        manager.popBackStackImmediate(null,
                FragmentManager.POP_BACK_STACK_INCLUSIVE);
        manager.beginTransaction()
                .replace(containerId, fragment)
                .commitAllowingStateLoss();

        if (fragment instanceof NavigationFragment) {
            ((NavigationFragment) fragment).setNavigationController(this);
        }

        update(oldFragment, fragment, Update.will);
        manager.executePendingTransactions();
        update(oldFragment, fragment, Update.did);

        setBackButtonHidden(true);
    }

    public void navigateTo(final Fragment fragment) {
        final Fragment oldFragment = currentFragment;
        if (android.os.Build.VERSION.SDK_INT > 21) {
            fragment.setEnterTransition(new Slide(Gravity.END));
            fragment.setExitTransition(new Slide(Gravity.START));
        }

        if (currentFragment == null) { return; }

        manager.executePendingTransactions();
        fragmentStack.push(fragment);
        resetClickable();
        manager.beginTransaction()
                .add(containerId, fragment)
                .addToBackStack(null)
                .commitAllowingStateLoss();

        if (fragment instanceof NavigationFragment) {
            ((NavigationFragment) fragment).setNavigationController(this);
        }

        update(oldFragment, fragment, Update.will);
        manager.executePendingTransactions();
        update(oldFragment, fragment, Update.did);

        setBackButtonHidden(false);
    }

    public boolean navigateBack() {
        final Fragment oldFragment = currentFragment;
        if (fragmentStack.size() > 0) {
            fragmentStack.pop();
        }

        Fragment fragment = topFragment();
        if (fragment == null || manager.getBackStackEntryCount() == 0) {
            return false;
        }

        manager.executePendingTransactions();
        update(oldFragment, fragment, Update.will);
        manager.popBackStackImmediate(null, 0);
        update(oldFragment, fragment, Update.did);

        setBackButtonHidden(fragmentStack.size() <= 1);
        return true;
    }

    public boolean onBackPressed() {
        Fragment fr = topFragment();
        if (fr instanceof OnBackPressedListener) {
            return ((OnBackPressedListener) fr).onBackPressedHandled();
        }
        else {
            return navigateBack();
        }
    }

    public boolean isPresenting() {
        return topFragment() != null;
    }

    @Nullable
    public Fragment topFragment() {
        if (fragmentStack.size() == 0) {
            return currentFragment;
        }
        return fragmentStack.peek();
    }

    private void resetClickable() {
        Fragment top = topFragment();
        for (Fragment s : fragmentStack) {
            boolean clickable = false;
            if (s == top) {
                clickable = true;
            }
            View v = s.getView();
            if (v != null) {
                v.setClickable(clickable);
            }
        }
    }

    // NAVIGATION BAR

    @Nullable
    public NavigationBar getNavigationBar() {
        return navigationBar;
    }

    public void setNavigationBar(NavigationBar navigationBar) {
        this.navigationBar = navigationBar;
        navigationBar.setDelegate(this);
        navigationBar.setBackButtonHidden(!shouldHideNavigationBar);
    }

    private void setBackButtonHidden(boolean value) {
        shouldHideNavigationBar = value;
        if (navigationBar != null) {
            navigationBar.setBackButtonHidden(value);
        }
    }

    // HELPER

    private void update(Fragment oldFragment, Fragment fragment, Update update) {
        if (!isVisible) { return; }

        if (update == Update.will) {
            if (fragment instanceof NavigationFragment) {
                ((NavigationFragment) fragment).viewWillAppear();
            }

            if (oldFragment instanceof NavigationFragment) {
                ((NavigationFragment) oldFragment).viewWillDisappear();
            }
        }
        else {
            if (fragment instanceof NavigationFragment) {
                ((NavigationFragment) fragment).viewDidAppear();
            }
            else {
                fragment.onResume();
            }

            if (oldFragment instanceof NavigationFragment) {
                ((NavigationFragment) oldFragment).viewDidDisappear();
            }
            else if (oldFragment != null) {
                oldFragment.onPause();
            }
        }
    }

    @Override public void navigationBarDidClickBackButton(@NotNull NavigationBar instance) {
        navigateBack();
    }
}

File: app/src/main/java/com/scruge/scruge/dependencies/navigation/NavigationController.kt
package com.scruge.scruge.dependencies.navigation

import android.transition.Slide
import android.view.Gravity

import com.scruge.scruge.view.views.NavigationBar

import java.util.Stack
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

class NavigationController(private val manager: FragmentManager, val containerId: Int) :
        NavigationBar.Delegate {

    /// return true if handled, false to pass further
    interface OnBackPressedListener {
        fun onBackPressedHandled(): Boolean
    }

    // FRAGMENT

    private var shouldHideNavigationBarBackButton = true
    private var fragmentStack = Stack<Fragment>()
    private var rootFragment: Fragment? = null

    var isVisible = true

    var navigationBar: NavigationBar? = null
        set(navigationBar) {
            field = navigationBar
            navigationBar?.delegate = this
            navigationBar?.isBackButtonHidden = shouldHideNavigationBarBackButton
            updateViews()
        }

    val topFragment: Fragment? get() = if (fragmentStack.size == 0) rootFragment else fragmentStack.peek()

    // METHODS

    fun replaceRoot(fragment: Fragment) {
        val oldFragment = rootFragment
        rootFragment = fragment

        manager.executePendingTransactions()
        fragmentStack = Stack()
        manager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        manager.beginTransaction().replace(containerId, fragment).commitAllowingStateLoss()
        manager.executePendingTransactions()

        if (fragment is NavigationFragment) {
            fragment.navigationController = this
        }

        update(oldFragment, fragment)
        updateViews()
    }

    fun popToRoot() {
        rootFragment?.let { replaceRoot(it) }
    }

    fun replaceTop(fragment: Fragment) {
        if (navigateBack()) {
            navigateTo(fragment)
        }
        else {
            replaceRoot(fragment)
        }
    }

    fun navigateTo(fragment: Fragment) {
        if (rootFragment == null) { return }

        val oldFragment = rootFragment
        if (android.os.Build.VERSION.SDK_INT > 21) {
            fragment.enterTransition = Slide(Gravity.END)
            fragment.exitTransition = Slide(Gravity.START)
        }

        manager.executePendingTransactions()
        fragmentStack.push(fragment)
        resetClickable()
        manager.beginTransaction().add(containerId, fragment).addToBackStack(null).commitAllowingStateLoss()
        manager.executePendingTransactions()

        if (fragment is NavigationFragment) {
            fragment.navigationController = this
        }

        update(oldFragment, fragment)
        updateViews()
    }

    fun navigateBack(): Boolean {
        if (fragmentStack.isEmpty()) { return false }

        val oldFragment = fragmentStack.pop()
        val fragment = topFragment!!

        manager.executePendingTransactions()
        manager.popBackStackImmediate(null, 0)
        manager.executePendingTransactions()

        if (fragment is NavigationFragment) {
            fragment.navigationController = this
        }

        update(oldFragment, fragment)
        updateViews()
        return true
    }

    fun onBackPressed(): Boolean {
        val value = (topFragment as? OnBackPressedListener)?.onBackPressedHandled() ?: return navigateBack()
        return if (!value) { navigateBack() } else { value }
    }

    fun updateViews() {
        setBackButtonHidden(fragmentStack.size == 0)
        (topFragment as? NavigationFragment)?.let {
            navigationBar
                    ?.setTitle(it.title)
                    ?.setupRightActionButton(it.navigationBarButton)
                    ?.setupRightActionButton2(it.navigationBarButton2)
                    ?.setHidden(it.shouldHideNavigationBar)
        }
    }

    // PRIVATE

    private fun resetClickable() {
        val top = topFragment
        for (fragment in fragmentStack) {
            var clickable = false
            if (fragment === top) {
                clickable = true
            }
            fragment.view?.isClickable = clickable
        }
    }

    private fun setBackButtonHidden(value: Boolean) {
        shouldHideNavigationBarBackButton = value
        navigationBar?.isBackButtonHidden = value
    }

    // HELPER

    private fun update(oldFragment: Fragment?, fragment: Fragment) {
        if (!isVisible) { return }

        (fragment as? NavigationFragment)?.viewDidAppear() ?: fragment.onResume()
        (oldFragment as? NavigationFragment)?.viewDidDisappear() ?: oldFragment?.onPause()
    }

    override fun navigationBarDidClickBackButton(instance: NavigationBar) {
        navigateBack()
    }
}
